### ISSUE TITLE
Apollo dev tools 4.0 not showing any data with multiple Apollo Clients

### ISSUE THREAD

**Issue by Rigel772** [apollographql/apollo-client-devtools#822]
**Intended outcome:**
I expect to be able to see the queries data in browser extension.

**Actual outcome:**
Not showing any data.

**How to reproduce the issue:**
Install extension in browser.

**Desktop (please complete the following information):**
 - OS: Linux Manjaro
 - Browser chromium, firefox
 - Browser version: Firefox Developer Edition 98, Chromium Version 98.0.4758.102 (Official Build) Manjaro Linux (64-bit)
 - Extension version 4.0

The dev tools version is 4, same situation in Chrome as in Firefox. "@apollo/client": "^3.5.7",
```
const createApolloClient = (authToken) => {
  return new ApolloClient({
    link: new HttpLink({
      uri: "https://...url",
      headers: {
        Authorization: `Bearer ${authToken}`,
      },
    }),
    cache: new InMemoryCache(),
    connectToDevTools: true, // should not be necessary but added just in case
    defaultOptions: {
      watchQuery: {
        fetchPolicy: "cache-and-network",
        pollInterval: 3000,
      },
    },
  });
};
```

The application is React + Vitejs, if it makes any difference... No idea where to start any debuging, please help...

**Comment by asgeo1** [comment]
I'm having the same issue with `4.0.0` extension in Chrome. I've tried the [troubleshooting steps](https://support.apollo.io/hc/en-us/articles/115003126012-Troubleshooting-the-Apollo-Chrome-Extension) but they didn't help.

The "Explorer" tab works fine, and executing any queries there will populate the Queries/Cache tabs.

However, executing queries in my app does nothing, so does not seem to be listening.

**Comment by asgeo1** [comment]
Actually, it's working for me now. I find that `connectToDevTools: true` is necessary, as well I found it didn't take effect immediately - I had to hard-refresh the page a couple of times.

**Comment by Rigel772** [comment]
Sadly does not work for me....

**Comment by henrinormak** [comment]
Had the same issue, but noticed that the code was creating multiple apollo clients - if this happens, only the first client is attached to the window as `__APOLLO_CLIENT__`, which is needed for the dev tools to work.

In short - make sure you only create a single apollo client, or if you do create multiple, make sure to assign the one you want to inspect via dev-tools to the window manually.

**Comment by dahei** [comment]
> Had the same issue, but noticed that the code was creating multiple apollo clients - if this happens, only the first client is attached to the window as `__APOLLO_CLIENT__`, which is needed for the dev tools to work.
> 
> In short - make sure you only create a single apollo client, or if you do create multiple, make sure to assign the one you want to inspect via dev-tools to the window manually.

This was the missing piece! It looks like you can only watch one Apollo client at a time. Thanks @henrinormak !

**Comment by jpvajda** [comment]
Just sharing this related stack overflow post that seems to be relating to this issue. https://stackoverflow.com/a/72116931/1880490

**Comment by Rigel772** [comment]
Have found that "GraphQL Network Inspector" Chrome extension is working nicely

**Comment by Rigel772** [comment]
...actually, I might misunderstood its purpose...
I played with it a bit more and I managed somehow to see my unauthenticated queries in "Explorer" tab and after executing a query IN THE EXTENSION I can see it on "Queries" tab... however I do not need another graphiQL app, I expected to see queries being made IN THE APP, to be able to monitor the results, errors etc
How does it work for you? How do you use it?

**Comment by Rigel772** [comment]
...and regarding multiple instances of Apollo Client, the client attached to window of my app has "2" in the name: "ApolloClient2"...
Does it mean there is "ApolloClient1" somewhere? How can I check how many clients are there?
I am creating the client in useData hook like this:
```
export default function useData() {
  const [token, setToken] = useRecoilState(Token);
   const [client, setClient] = useState({});
  useEffect(() => {
    if (token) {
      setClient(() => createApolloClient(token.value));
    }
  }, [token]);
   return {
    client,
    DataProvider: ApolloProvider,
    setToken,
  };
}
```

**Comment by cheapsteak** [comment]
Hi there, 

How is useData used? How many places is it used?  
I might recommend trying a context provider approach (and use useMemo instead of useEffect) to see if that makes a difference. See here for an example: https://github.com/ellanan/mealkit/blob/main/src/ApolloClientProvider.tsx

But before you rewrite anything, could you please help us isolate where the problem might be by going to another site that uses apollo and see whether you see data there? e.g. https://studio.apollographql.com/login should show some queries  

<img width="1002" alt="image" src="https://user-images.githubusercontent.com/743976/176222845-02d76d6c-e604-46e6-a45d-f1f02079b7e8.png">

If you do see data there then we'll know that it's something to do with which instance of apollo client the extension is talking to

**Comment by Oikio** [comment]
Same problem here, did not work locally, but worked on https://studio.apollographql.com/login. 

Restarting Chrome helped.

**Comment by velislavgerov** [comment]
Had a similar issue while doing local development. The extension didn't work properly when I was connecting via `localhost`. Swapping it for `127.0.0.1` solved the issue for me.

**Comment by shloak17107** [comment]
> Had the same issue, but noticed that the code was creating multiple apollo clients - if this happens, only the first client is attached to the window as `__APOLLO_CLIENT__`, which is needed for the dev tools to work.
> 
> In short - make sure you only create a single apollo client, or if you do create multiple, make sure to assign the one you want to inspect via dev-tools to the window manually.

Hey, how do you check if there are indeed multiple instances of Apollo Client being created? And how do you assign the one you want to be able to inspect in Apollo Devtools?

**Comment by DaSchTour** [comment]
I have the same issue even with just one client. I removed all named clients. It's very frustrating not have not working dev tools. Sometime I also have some ghost data from ancient times. So it looks like sometime it works and then breaks and displays old data.

**Comment by jpvajda** [comment]
@MrDoomBringer This bug has been outstanding for some time, given your recent exploration into the duplicate queries issue we recently resolved, I was curious if you want to take a look that this one as well. 🙏  cc @bignimbus

**Comment by jpvajda** [comment]
@MrDoomBringer  @bignimbus  This might be an interesting  issue to pick up soon for investigation. What do you think?

**Comment by dcorb** [comment]
@MrDoomBringer  let me help you here: 
A reproducible test case. 
During the first 10 seconds, Apollo DevTools work, you can see the Queries for example.
https://90g9kq.csb.app/
After 10 seconds, a new ApolloClient is instantiated.
Full Codesandbox https://codesandbox.io/s/adoring-ishizaka-90g9kq?file=/src/index.js

**Comment by johnhok** [comment]
Recently ran into this issue due to late creation of our apollo client instance. There are also multiple instances across different teams which made utilizing dev tools almost impossible. 

I was able to fix it by cloning the extension, increasing the timeout it waits for the client to be instantiated and binded dev tools to a unique global variable here in the code https://github.com/apollographql/apollo-client-devtools/blob/1cfa015c109201139fd71fe6089d05d8d2505c8f/src/extension/tab/hook.ts#L221

Seems like a more user friendly option would be to provide some configurability and let users bind to a custom variable or to manually execute the binding.

**Comment by jerelmiller** [comment]
Hey @johnhok 👋 

We are considering a different approach to registering clients with the dev tools at some point. Unfortunately the way the Apollo Client and dev tools communicates is through a single `window.__APOLLO_CLIENT__` variable, which doesn't work super well with multiple clients. I know the current situation isn't perfect, but we are definitely thinking of ways to make it better.

No guarantees on timeline, but I can at least say we are thinking about it! I think it would be super helpful to be able to toggle between the two clients and inspect what each is doing independently.

If your fork is working for you now, I'd suggest using that for the time being. I'd prefer not to do much to the current `initializeDevtoolsHook` until we come up with a better way of registering clients. Hope that helps!

**Comment by kjhuang-db** [comment]
hi @johnhok did you end up forking the extension and install it from chrome?

**Comment by johnhok** [comment]
> hi @johnhok did you end up forking the extension, build it locally then load it as unpacked chrome extension?

Yep that's exactly what I did. I forked it and did the hacky modifications I mentioned about increasing the timeout and binding to a different global variable. Only the specific apollo client I was interested in gets assigned to that variable.

Once you build it with the instructions in the readme you can load it as an unpacked extension like you mentioned.

**Comment by kjhuang-db** [comment]
@johnhok thank you I plan to do so :), based on the [code](https://github.com/apollographql/apollo-client/blob/v3.6.9/src/core/ApolloClient.ts#L189), it does not seems a "different" global variable is needed though. can just suppress unwanted ones with explicit `connectToDevTools: false`, and leave only one with `connectToDevTools: true`. Then the global singleton `__APOLLO_CLIENT__` should be unique

**Comment by diogocapela** [comment]
Same issue here. Only one `ApolloClient` instance is being created and `window. __APOLLO_CLIENT__` is present.

Apollo Client Devtools does not show any data using `localhost` or `127.0.0.1`.

It does work when using the [GraphQL Network Inspector](https://chromewebstore.google.com/detail/ndlbedplllcgconngcnfmkadhokfaaln) extension, but that does not allow for cache inspection.

Any news on this issue?

**Comment by diogocapela** [comment]
Same issue here. Only one `ApolloClient` instance is being created, `connectToDevTools: true` is passed to the client, and `window.__APOLLO_CLIENT__` is present.

Apollo Client Devtools does not show any data using `localhost` or `127.0.0.1`.

It does work when using the [GraphQL Network Inspector](https://chromewebstore.google.com/detail/ndlbedplllcgconngcnfmkadhokfaaln) extension, but that does not allow for cache inspection.

Any news on this?

**Comment by jerelmiller** [comment]
Hey @diogocapela we've done a lot of work recently on the message passing between the client and devtools. Would you give a more recent version of the devtools a shot to see if it has resolved your issue finding the client?

**Comment by adrianwix** [comment]
For me it does not work in 4.9.0
Only one ApolloClient instance is being created, connectToDevTools: true is passed to the client, and window.__APOLLO_CLIENT__ is present. (As diogocapela mentioned)
@jerelmiller 

![image](https://github.com/apollographql/apollo-client-devtools/assets/37629556/3a3062b5-615a-4952-b46f-ff85b0de31c5)
![image](https://github.com/apollographql/apollo-client-devtools/assets/37629556/e3aaa952-06ac-48c7-84be-f8fcf8b97018)

**Comment by alessbell** [comment]
@adrianwix @diogocapela which version of Apollo Client are you using? Sorry to hear you're blocked on using the extension - as @jerelmiller said, we'd love to get some more info and debug further.

Since only one instance of the client is being created, please feel free to open a new issue. And if you're able to create a reproduction in an app we can run locally, that would be tremendously helpful 🙏

**Comment by adrianwix** [comment]
"@apollo/client": "^3.8.7" @alessbell. 
I came here because I also had this issue before but then I removed the multiple instances being create and realize that it was still not working, so not sure exactly where the problem is now. Previously it would be solved after removing the duplicated instances

**Comment by jerelmiller** [comment]
@adrianwix do you mind creating a separate issue for this? I'd like to track this separately and keep the discussion in that issue so we don't clutter up this one too much 🙂

**Comment by shamseer-ahammed** [comment]
> Had the same issue, but noticed that the code was creating multiple apollo clients - if this happens, only the first client is attached to the window as `__APOLLO_CLIENT__`, which is needed for the dev tools to work.
> 
> In short - make sure you only create a single apollo client, or if you do create multiple, make sure to assign the one you want to inspect via dev-tools to the window manually.

"make sure to assign the one you want to inspect via dev-tools to the window manually" how do you do that ?

**Comment by shamseer-ahammed** [comment]
i dont have duplicate instances and i am also getting the same, connectToDevTools: true
devtools shows up but wont show any data

chrome : Version 124.0.6367.119 (Official Build) (64-bit)
devttols version : 4.11.5

**Comment by kjhuang-db** [comment]
is this issue addressed?

**Comment by jerelmiller** [comment]
@kjhuang-db we've done a lot of work to shore up the message passing so I think its more stable than it used to be. Glad to hear its working better now! 

What is currently missing is the ability to switch between which client's data you're looking at in the devtools. Right now it only shows you the last one that connected to the devtools. I've got a [WIP PR](https://github.com/apollographql/apollo-client-devtools/pull/1418) that will be addressing this bit so that if it detects multiple clients, you'll be able to switch between them. I'm hoping to finish this up within the next week or two.

**Comment by jerelmiller** [comment]
Hey all 👋 

Thanks so much for the patience on this issue! There has been a lot of activity here and judging by the comments has caused quite a few headaches in the past.

I've got https://github.com/apollographql/apollo-client-devtools/pull/1418 ready to go that will fully enable this capability and allow you to connect and inspect multiple client instances in the devtools. This change will also take advantage of an upcoming feature in Apollo Client 3.11 that will allow you to [name your client](https://github.com/apollographql/apollo-client/pull/11936) for easier debugging in the devtools. Our plan is to try and release this change alongside 3.11 which should land some time next week (barring any last minute issues we find in that release).

I'm really hoping this will benefit apps that use multiple clients. I'm very excited to get this out there! Thanks again for your patience!!

**Comment by leosoaivan** [comment]
Any updates on the client's v3.11 release?

Currently facing issues with the messaging where the queries, mutations, and cache values are empty. This is with:
- Chrome 126.0.6478.183
- Apollo Client v3.10.8
- Apollo Devtools v4.16.1

In a production build, I noticed that our app's initial queries were caught by the devtools, before clearing out from the tool after a second or two, without a trace.

**Comment by jerelmiller** [comment]
@leosoaivan we're looking to get 3.11 out by latest tomorrow morning so expect it soon. We'll be pushing #1418 along with this to take advantage of the new devtools option in 3.11. 

The reason you're likely not seeing data is because any client instantiated after the first one will "override" the client the devtools is looking at. If the 2nd instance of the client has no queries, you won't see data available. I'm definitely looking forward to getting #1418 out to fix this issue 🙂

**Comment by leosoaivan** [comment]
@jerelmiller Appreciate the update.

I was just about to say that I was fairly certain we weren't initializing additional clients after double-checking, but I _just_ found another one in an area owned by another team 😩. 

Thanks again 😆.

**Comment by jerelmiller** [comment]
Apollo Client 3.11 has now been released alongside #1418 providing the capability to connect and inspect multiple client instances. If you find any issues with the new feature, please open a new issue and I'd be happy to take a look. Thanks!
